lib/context/provider.ts

```diff
@@ -0,0 +1,113 @@
+import type { ReactiveController, ReactiveElement } from 'lit';
+import type { Context, ContextCallback, ContextEvent, UnknownContext } from './event.js';
+import type { ColorContextProviderOptions } from './decorators.js';
+import type { ColorTheme } from './types.js';
+
+import { contextEvents, ColorContextController } from './controller.js';
+
+import { Logger } from '@patternfly/pfe-core/controllers/logger.js';
+
+/**
+ * `ColorContextProvider` is responsible to derive a context value from CSS and provide it to its
+ * descendents.
+ */
+export class ColorContextProvider<T extends ReactiveElement> extends ColorContextController<T> implements ReactiveController {
+  #attribute: string;
+
+  /** Cache of context callbacks. Call each to update consumers */
+  #callbacks = new Set<ContextCallback<ColorTheme|null>>();
+
+  /** Mutation observer which updates consumers when `on` or `color-palette` attributes change. */
+  #mo = new MutationObserver(() => this.update(this.#contextVariable));
+
+  /**
+     * Cached (live) computed style declaration
+     * @see https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle
+     */
+  #style: CSSStyleDeclaration;
+
+  protected logger: Logger;
+
+  #initialized = false;
+
+  /** Return the current CSS `--context` value, or null */
+  get #contextVariable(): ColorTheme | null {
+    return this.#style.getPropertyValue('--context').trim() as ColorTheme || null;
+  }
+
+  constructor(host: T, options?: ColorContextProviderOptions<T>) {
+    const { attribute = 'color-palette', ...rest } = options ?? {};
+    super(host, rest);
+    this.logger = new Logger(host);
```

Comment:
maybe adding a private-to-the-provider ColorContextConsumer here would do the trick